packages/discord-request/src/manager.ts

```diff
@@ -0,0 +1,330 @@
+import debug from "debug";
+import { OFFSET, ONE_DAY, ONE_SECOND, sleep } from "./util/time";
+import { getRouteInformation, getRouteKey } from "./util/routes";
+import Queue from "./queue";
+import { RateLimitData, RequestData, RequestOptions, Route } from "./types";
+
+const log = debug("discord-request:manager");
+
+type Bucket = {
+	key: string;
+	lastRequest: number;
+};
+
+export type Config = {
+	api: string;
+	cdn: string;
+	headers: Record<string, string>;
+	retries: number;
+	timeout: number;
+	userAgent: string;
+	version: number;
+	globalRequestsPerSecond: number;
+};
+
+export type Callbacks = {
+	onBucketSweep?: (swept: Map<string, Bucket>) => void;
+	onQueueSweep?: (swept: Map<string, Queue>) => void;
+	onRateLimit?: (data: RateLimitData) => void;
+	onRequest?: (
+		parameters: Route,
+		resource: string,
+		init: RequestInit,
+		retries: number
+	) => void;
```

Comment:
This is one of those places where we're gonna need to explicitly add | undefined after the other PR FYI